src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,137 @@
+// <copyright file="AWSMessagingUtils.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.Lambda.SNSEvents;
+using Amazon.Lambda.SQSEvents;
+using Newtonsoft.Json;
+using OpenTelemetry.Context.Propagation;
+
+namespace OpenTelemetry.Instrumentation.AWSLambda.Implementation;
+
+internal class AWSMessagingUtils
+{
+    // SNS attribute types: https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+    private const string SnsAttributeTypeString = "String";
+    private const string SnsAttributeTypeStringArray = "String.Array";
+    private const string SnsMessageAttributes = "MessageAttributes";
+
+    internal static PropagationContext ExtractParentContext(SQSEvent sqsEvent)
+    {
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
```

Comment:
Shouldn't the parent be the trace context created by the Lambda function?

Reply:
This is the parent *of* the trace context created by the Lambda function.

Reply:
It looks like we are using the span context in the last message's message attribute as a parent.

What happens in a batch case? Say we have multiple messages in a Lambda event. We may not want to use the last message's upstream trace context.

Reply:
I don't think using the last record is the right approach, given each message in the queue could have come from a different source with a completely different original trace. Using the last would mean all other traces wouldn't be linked/used.